datalad_revolution/metadata/extractors/runprov.py

```diff
@@ -0,0 +1,106 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Metadata extractor for provenance information in DataLad's `run` records
+
+Concept
+-------
+
+- Find all the commits with a run-record encoded in them
+- the commit SHA provides @id for the "activity"
+- pull out the author/date info for annotation purposes
+- pull out the run record (at the very least to report it straight
+  up, but there can be more analysis of the input/output specs in
+  the context of the repo state at that point)
+- pull out the diff: this gives us the filenames and shasums of
+  everything touched by the "activity". This info can then be used
+  to look up which file was created by which activity and report
+  that in the content metadata
+"""
+
+
+from .base import MetadataExtractor
+from datalad.support.json_py import loads as jsonloads
+
+import logging
+lgr = logging.getLogger('datalad.metadata.extractors.runprov')
+
+
+class RunProvenanceExtractor(MetadataExtractor):
+    def __call__(self, dataset, process_type, status):
+        # shortcut
+        ds = dataset
+
+        stdout, stderr = ds.repo._git_custom_command(
+            None,
+            ['git', 'log', '-F',
+             '--grep', '=== Do not change lines below ===',
+             "--pretty=tformat:%x00%x00record%x00%n%H%x00%aN%x00%aE%x00%aI%n%B%x00%x00diff%x00",
+             "--raw", "--no-abbrev",
+            ]
+        )
```

Comment:
This parser looks nice.  `rerun` could take some inspiration, as noted in datalad/datalad#3372.

Reply:
@kyleam I RF'ed the parser into a separate function to facilitate a later RF/move.